File: src/main/java/kitchenpos/order/tobe/eatinorder/application/EatInOrderService.java
package kitchenpos.order.tobe.eatinorder.application;

import kitchenpos.common.event.publisher.OrderTableClearEvent;
import kitchenpos.order.tobe.eatinorder.application.dto.request.EatInOrderCreateRequest;
import kitchenpos.order.tobe.eatinorder.domain.EatInOrder;
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderLineItem;
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderRepository;
import kitchenpos.order.tobe.eatinorder.domain.OrderLineItems;
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTable;
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableRepository;
import kitchenpos.order.tobe.eatinorder.domain.validate.MenuValidator;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

@Service
public class EatInOrderService {
    private final EatInOrderRepository eatInOrderRepository;
    private final OrderTableRepository orderTableRepository;
    private final MenuValidator menuValidator;
    private final ApplicationEventPublisher eventPublisher;

    public EatInOrderService(
            final EatInOrderRepository eatInOrderRepository,
            final OrderTableRepository orderTableRepository,
            final MenuValidator menuValidator,
            final ApplicationEventPublisher eventPublisher
    ) {
        this.eatInOrderRepository = eatInOrderRepository;
        this.orderTableRepository = orderTableRepository;
        this.menuValidator = menuValidator;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    public EatInOrder create(final EatInOrderCreateRequest request) {
        final List<EatInOrderLineItem> orderLineItems = request.orderLineItems().stream()
                .map(dto -> new EatInOrderLineItem(dto.menuId(), dto.quantity(), dto.price()))
                .toList();

        OrderLineItems validatedOrderLineItems = new OrderLineItems(orderLineItems, menuValidator);
        OrderTable orderTable = findOrderTableById(request.orderTableId());

        EatInOrder order = new EatInOrder(UUID.randomUUID(), validatedOrderLineItems, orderTable);
        return eatInOrderRepository.save(order);
    }

    @Transactional
    public EatInOrder accept(final UUID orderId) {
        EatInOrder order = findOrderById(orderId);

        order.accept();

        return order;
    }

    @Transactional
    public EatInOrder serve(final UUID orderId) {
        EatInOrder order = findOrderById(orderId);

        order.serve();

        return order;
    }

    @Transactional
    public EatInOrder complete(final UUID orderId) {
        EatInOrder order = findOrderById(orderId);

        order.complete();

        eventPublisher.publishEvent(new OrderTableClearEvent(order.getOrderTable().getId()));

        return order;
    }

    @Transactional(readOnly = true)
    public List<EatInOrder> findAll() {
        return eatInOrderRepository.findAll();
    }

    private OrderTable findOrderTableById(UUID orderTableId) {
        return orderTableRepository.findById(orderTableId)
                .orElseThrow(NoSuchElementException::new);
    }

    private EatInOrder findOrderById(UUID orderId) {
        return eatInOrderRepository.findById(orderId)
                .orElseThrow(NoSuchElementException::new);
    }

}

File: src/main/kotlin/kitchenpos/order/tobe/eatinorder/application/EatInOrderService.kt
package kitchenpos.order.tobe.eatinorder.application

import java.util.*
import kitchenpos.order.tobe.common.OrderMenuClient
import kitchenpos.order.tobe.eatinorder.application.dto.CreateEatInOrderReq
import kitchenpos.order.tobe.eatinorder.domain.EatInOrder
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderLineItem
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderLineItems
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderRepository
import kitchenpos.order.tobe.eatinorder.domain.OrderTableRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service("tobeEatInOrderService")
class EatInOrderService(
    private val eatInOrderRepository: EatInOrderRepository,
    private val orderTableRepository: OrderTableRepository,
    private val orderMenuClient: OrderMenuClient,
) {

    @Transactional
    fun create(req: CreateEatInOrderReq): UUID {
        val orderLineItems = req.orderLineItems.map { orderLineItem ->
            EatInOrderLineItem.create(
                seq = orderLineItem.seq,
                menuInfo = orderMenuClient.getMenu(orderLineItem.menuId),
                quantity = orderLineItem.quantity,
            )
        }
        val orderTable = orderTableRepository.findById(req.orderTableId).orElseThrow {
            NoSuchElementException("주문 테이블이 존재하지 않습니다.")
        }
        val eatInOrder = EatInOrder.create(
            orderTable = orderTable,
            orderLineItems = EatInOrderLineItems(orderLineItems),
        )
        return eatInOrderRepository.save(eatInOrder).id
    }

    @Transactional
    fun accept(orderId: UUID) {
        val eatInOrder = eatInOrderRepository.findById(orderId).orElseThrow {
            NoSuchElementException("주문이 존재하지 않습니다.")
        }
        eatInOrder.accept()
    }

    @Transactional
    fun serve(orderId: UUID) {
        val eatInOrder = eatInOrderRepository.findById(orderId).orElseThrow {
            NoSuchElementException("주문이 존재하지 않습니다.")
        }
        eatInOrder.serve()
    }

    @Transactional
    fun complete(orderId: UUID) {
        val eatInOrder = eatInOrderRepository.findById(orderId).orElseThrow {
            NoSuchElementException("주문이 존재하지 않습니다.")
        }
        eatInOrder.complete()
        eatInOrderRepository.save(eatInOrder)
    }
}
